180 check creation of composition formula

JFRudzinski opened the pull request:
## Summary of Changes

* **Fix: `Optional` NameError**
  Don’t import `Optional` only inside `TYPE_CHECKING`. Use a runtime import (`from typing import Optional`) or string annotations (e.g., `'Optional[str]'`).

* **`resolve_composition_formula` Overhaul**

  * **Internal nodes:** build composition by counting child `branch_label`s (use `"Unknown"` if missing).
  * **Leaves:** derive composition from root `symbols` sliced by this node’s `particle_indices`; skip if indices/labels are missing.
  * **Honor custom formulas:** never overwrite if `composition_formula` is already set.
  * Implemented via a simple recursive `walk(system_parent)`.

* **`is_representative` Temporarily Ignored for Composition**
  Composition formulas are computed regardless of `is_representative` until scope is finalized. Tests updated to expect formulas even when not representative.

* **Stronger `ModelSystem.symbols`**

  * **Root:** collect labels from `AtomsState` / `CGBeadState` with sensible fallbacks.
  * **Child:** slice from the **ultimate root**; validate indices:

    * must be integers, **non-negative**, and **in range** → otherwise return `[]`.
  * Establishes explicit **“no negative indices”** policy (replacing previous xfail assumption).

* **Bond-List Utilities**
  `get_bond_list(set_local=True)` now **persists** filtered bonds on the child (`child.bond_list`), i.e., order is preserved, and returns an empty `(0, 2)` array when appropriate.

* **Hierarchy Helpers**
  `is_root_system()` / `get_root_system()` use **`m_def` identity checks** (robust to proxies) instead of `isinstance`.

* **ASE Conversions**
  `to_ase_atoms()` **rejects non-element symbols** for atomic systems with clear logging.
  `from_ase_atoms()` populates positions, cell/PBC, and recreates particle states from the ASE object.

## Review Roles

- @ladinesa please provide a technical coding review, I am especially unsure about the dealing with arrays vs. lists and associated typing in the functions where I am slicing root quantities with the subsystem particle_indices

- @EBB2675 please provide a review of the new/adjusted functionalities: see `ModelSystem.resolve_composition_formula()`, `ModelSystem.symbols`, `ModelSystem.is_root_system()`, `ModelSystem.get_root_system()`, `utils.get_composition()`

## Notes

- Passes nomad-parser-plugins-simulation tests

JFRudzinski committed:
a mess of preliminary changes, including new implementation of ModelSystem.symbols, get_bond_list, get_root, etc

JFRudzinski committed:
working tests

JFRudzinski committed:
clean comments

JFRudzinski committed:
type annotation for idx

JFRudzinski committed:
ruff

JFRudzinski committed:
idx type condition fix

JFRudzinski commented on src/nomad_simulations/schema_packages/utils/utils.py:
```diff
@@ -1,14 +1,16 @@
 
-if TYPE_CHECKING:
-    from typing import Callable, Optional
+# if TYPE_CHECKING:
```
@ladinesa do we still need this conditional import here? I have the feeling like this was some previous issue that is now solved.

ladinesa replied:
yes you can remove this, i think i dod the same for the parsers repo, unless there is a cyclic import error which i doubt since there are third party packages.

JFRudzinski replied:
ok great, will address it for the other cases in #218

ladinesa commented on src/nomad_simulations/schema_packages/general.py:
```diff
@@ -207,97 +207,57 @@ def _set_system_branch_depth(
-        """Determine and set the composition formula for `system_parent` and all of its
-        descendants.
-
-        Args:
```
insert rules but do not remove args

ladinesa commented on src/nomad_simulations/schema_packages/general.py:
```diff
@@ -207,97 +207,57 @@ def _set_system_branch_depth(
-                    if subsystem.branch_label is not None
-                    else 'Unknown'
-                    for subsystem in subsystems
+        def set_for(system: ModelSystem) -> None:
```
i suggest keeping the old name, or more explicit `set_formula`

ladinesa commented on src/nomad_simulations/schema_packages/general.py:
```diff
@@ -207,97 +207,57 @@ def _set_system_branch_depth(
-                    children_names=subsystem_labels
+                    children_names=children_names
                 )
+                return
```
get rid these return l238, 242 and assign formula at the end, if they can be extracted from the various cases

JFRudzinski replied:
There are not overlapping cases, Leaf nodes are set one way always, internal nodes set the other way always

ladinesa commented on src/nomad_simulations/schema_packages/general.py:
```diff
@@ -207,97 +207,57 @@ def _set_system_branch_depth(
+
+        def walk(node: ModelSystem) -> None:
+            set_for(node)
+            for child in node.sub_systems or []:
```
sub-sections default to [], no need for `or []`

ladinesa commented on src/nomad_simulations/schema_packages/general.py:
```diff
@@ -207,97 +207,57 @@ def _set_system_branch_depth(
+                        children_names=['Unknown'] * n
+                    )
+
+        def walk(node: ModelSystem) -> None:
```
this is essentially already `self.resolve_composition_formula`, so no need to define walk just call `resolve_composition_formula` iteratively. this should actually be a static method as it appears it does not depend on the object i.e. no `self.` calls in the function.

JFRudzinski replied:
is it better to use @staticmethod or move outside of Simulation()?

ladinesa replied:
these are specific to the class so they should belong to it i suppose.

ladinesa commented on src/nomad_simulations/schema_packages/general.py:
```diff
@@ -320,8 +280,9 @@ def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
 
-            if is_not_representative(model_system=system_parent, logger=logger):
-                continue
+            # ! Address when we know the role of is_representative
```
add TODO

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1054,28 +1055,64 @@ def _is_atomic(self, value: Optional[bool]) -> None:
 
-    # ? Is it better to get symbols with logging or make a property without?
-    def get_symbols(self, logger: 'BoundLogger') -> list[str]:
+    @property
```
for now, let it stay as is, do not use property in MSection as it  technically, all m-quantities are properties. i will add log decorator after

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1054,28 +1055,64 @@ def _is_atomic(self, value: Optional[bool]) -> None:
-    # ? Is it better to get symbols with logging or make a property without?
-    def get_symbols(self, logger: 'BoundLogger') -> list[str]:
+    @property
+    def symbols(self) -> list[str]:
```
rename this to what it does as a method, i.e. `get_symbols` to make it consistent with the others

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1054,28 +1055,64 @@ def _is_atomic(self, value: Optional[bool]) -> None:
-            if not symbol:
-                logger.warning('missing symbol in ParticleState.')
+
+        def _root_symbols(ms: 'ModelSystem') -> list[str]:
```
you can remove this method and put everything under a static method get_symbols that take in a model_system. As much as possible vall this only once and any self.symbols ref should ve made to the msection quantity which ia derived from this method.

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1054,28 +1055,64 @@ def _is_atomic(self, value: Optional[bool]) -> None:
+            symbols: list[str] = []
+            for particle_state in ms.particle_states:
+                symbol = None
+                if isinstance(particle_state, AtomsState):
```
ia it possible to handle this logic in the PartcleState classes so you do not have the if statements for various types

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1054,28 +1055,64 @@ def _is_atomic(self, value: Optional[bool]) -> None:
+                    if not symbol:
+                        alts = particle_state.alt_labels
+                        if alts:
+                            try:
```
the try is a bit unnecessary  symbol = alt_labels(0) if al_labels else None

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1054,28 +1055,64 @@ def _is_atomic(self, value: Optional[bool]) -> None:
+                else:
+                    symbol = particle_state.label
+                if not symbol:
+                    return []
```
maybe put a assign symbols to empty list then break rather than return

ladinesa replied:
maybe put a assign symbols to empty list then break rather than return, several returns are harder to maintain.

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1369,17 +1411,41 @@ def is_ge_structure(self, other: 'ModelSystem') -> bool:
 
-    # functions for traversing the ModelSystem hierarchy
-    def get_root_system(self) -> 'ModelSystem':
+    def is_root_system(self) -> bool:
```
maybe a one-liner if parent is None or parent
m_def is not ModelSystem.m_def?

ladinesa replied:
is the old method not used anymore? ok it follows

ladinesa commented on src/nomad_simulations/schema_packages/utils/utils.py:
```diff
@@ -123,14 +119,48 @@ def is_not_representative(model_system, logger: 'BoundLogger'):
+#     children_count_tup = np.unique(children_names, return_counts=True)
+#     formula = ''.join([f'{name}({count})' for name, count in zip(*children_count_tup)])
+#     return formula if formula else None
+def get_composition(children_names: list[str]) -> Optional[str]:
```
is there not an ase method that does this? or maybe try to base it on some ase functions if the children names are themselves formulas

JFRudzinski replied:
This is a generalization of what ase does, I looked at their code but I didn't take away anything useful for now, it depends a little if people like this convention I developed or not

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1097,7 +1134,7 @@ def are_valid_chemical_symbols(self, logger: 'BoundLogger') -> bool:
             bool: True if all chemical symbols are valid, False otherwise.
         """
-        symbols = self.get_symbols(logger)
+        symbols = self.symbols
```
please see my previous if possible do not call the method more than once and rely on the msection quantity derived from the get_symbols method.

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1369,17 +1411,41 @@ def is_ge_structure(self, other: 'ModelSystem') -> bool:
         system = self
-        while isinstance(system.m_parent, ModelSystem):
-            system = system.m_parent
+        while True:
```
use the `is_root_system` above instead to break loop

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1369,17 +1411,41 @@ def is_ge_structure(self, other: 'ModelSystem') -> bool:
+            break
         return system
 
     # functions for working with molecules
```
i think this can also be a static method, remove the set_local arg, setting quantities should be done elsewhere.

JFRudzinski replied:
I did improve this I think but I think get_bond_list() should be a class method, like get_symbols(), otherwise the implementation in terms of calling the function feels a bit strange, because I am always doing `ModelSystem.get_bond_list(self)`, please check again

ladinesa replied:
yes i agree

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1395,27 +1461,29 @@ def get_bond_list(self, set_local: bool = False) -> np.ndarray:
+            return empty_return
+
+        idx: NDArray[np.int32]
+        idx = (
```
in line typing idx: np.ndarray, we did have some issues with NDArray? mayb use np.ndarray

ladinesa commented on src/nomad_simulations/schema_packages/utils/utils.py:
```diff
@@ -123,14 +119,48 @@ def is_not_representative(model_system, logger: 'BoundLogger'):
 
 # TODO remove function in nomad.atomutils
-def get_composition(children_names: 'list[str]') -> str:
+# def get_composition(children_names: 'list[str]') -> str:
```
remove comments

ladinesa commented on src/nomad_simulations/schema_packages/utils/utils.py:
```diff
@@ -123,14 +119,48 @@ def is_not_representative(model_system, logger: 'BoundLogger'):
+    # Ensure string dtype; np.unique returns sorted unique names and counts
+    arr = np.asarray(children_names, dtype=object).astype(str)
+    names, counts = np.unique(arr, return_counts=True)  # case-sensitive
+    # ? switch the above line for case insensitive sorting below?
```
put a TODO if unresolvd.

ladinesa commented on tests/test_general.py:
```diff
@@ -98,15 +98,30 @@ def get_flat_depths(system_parent: ModelSystem, quantity_name: str, value=None):
                 [None, None, None, None, None],
             ),  # pure system
-            (
+            # (
```
remove comments

ladinesa commented on src/nomad_simulations/schema_packages/utils/utils.py:
```diff
@@ -123,14 +119,48 @@ def is_not_representative(model_system, logger: 'BoundLogger'):
+#     children_count_tup = np.unique(children_names, return_counts=True)
+#     formula = ''.join([f'{name}({count})' for name, count in zip(*children_count_tup)])
+#     return formula if formula else None
+def get_composition(children_names: list[str]) -> Optional[str]:
```
not sure if optional is still a thing or str | None

ladinesa reviewed (changes requested):
The get_ functions should be made consistent  and I believe should be static methods. Thee are are simplifications that are necessary. The tests look reasonable although have not checked the coverage.

EBB2675 commented on src/nomad_simulations/schema_packages/utils/utils.py:
```diff
@@ -123,14 +119,48 @@ def is_not_representative(model_system, logger: 'BoundLogger'):
+#     children_count_tup = np.unique(children_names, return_counts=True)
+#     formula = ''.join([f'{name}({count})' for name, count in zip(*children_count_tup)])
+#     return formula if formula else None
+def get_composition(children_names: list[str]) -> Optional[str]:
```
in `get_composition()`  without sorting the keys the string order may differ across runs. This can make outputs non-deterministic. I don't really know to what extent it might be problematic though. Please see my randomized test:

```
Results over 20 runs:
01: H3 Unknown h Na+2
02: H3 h Na+2 Unknown
03: Unknown Na+2 H3 h
04: Unknown h Na+2 H3
05: H3 h Na+2 Unknown
06: Unknown h H3 Na+2
07: Unknown Na+2 H3 h
08: Unknown Na+2 H3 h
09: H3 Na+2 Unknown h
10: Na+2 Unknown h H3
11: H3 Unknown Na+2 h
12: Na+2 Unknown h H3
13: H3 Unknown Na+2 h
14: H3 Unknown Na+2 h
15: Na+2 H3 Unknown h
16: Na+2 H3 h Unknown
17: Unknown h Na+2 H3
18: H3 h Unknown Na+2
19: h H3 Na+2 Unknown
20: H3 Na+2 h Unknown

```

JFRudzinski replied:
I think I fixed this, could you please check again?

EBB2675 commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1369,17 +1411,41 @@ def is_ge_structure(self, other: 'ModelSystem') -> bool:
-            ModelSystem: The top-level (root) ModelSystem.
+        return parent.m_def is not ModelSystem.m_def
+
+    def get_root_system(self) -> 'ModelSystem':
```
would it make sense to add a simple visited-set or max-depth guard for edge cases? 

```
a = ModelSystem(branch_label="A")
b = ModelSystem(branch_label="B", m_parent=a)
c = ModelSystem(branch_label="C", m_parent=b)

# i create a cycle: A points back to C
a.m_parent = c

# aaaand infinite loop
c.get_root_system()
```

JFRudzinski replied:
Will address in #244

EBB2675 commented on src/nomad_simulations/schema_packages/general.py:
```diff
@@ -207,97 +207,57 @@ def _set_system_branch_depth(
             )
 
     def resolve_composition_formula(self, system_parent: ModelSystem) -> None:
```
I wish to check my understanding

- **in the case of internal node:** if a node has `sub_systems`, the formula is built via children's `branch_label`.
- **in the case of a leaf node:** if a node has `particle_indices`, then the formula is built from the root

this is great if i have only `sub_systems` or only `particle_indices`

do we need to cover the cases that a node has both? Because right now the function would silently choose one path

EBB2675 replied:
maybe at least we can make a note of the behavior in the docstring:

_If both are present, `sub_systems` take precedence_ or sth like that

JFRudzinski replied:
There are "Rules" already in the docstring, can you tell me more precisely if these are not clear?

In case of internal nodes (with children) but no branch_labels, should we revert to using the particle labels instead of `Unknown(XXX)`?

JFRudzinski replied:
I changed the rules now, to also use particle_labels if all branch_labels are Unknown, please check the clarity / if you agree

JFRudzinski committed:
redo of resolve_comp_formula & symbols from review, still messy but tests work

JFRudzinski committed:
cleaned composition formula, added new functionality for missing branch labels,  and fixed typing issue in tests, updated test results

JFRudzinski committed:
refactor get_symbols

JFRudzinski committed:
is_root_system simplified

JFRudzinski committed:
get_root_system simplified

JFRudzinski committed:
bond_list is getter only

JFRudzinski committed:
robust sorting in get_composition

JFRudzinski committed:
fix util imports

JFRudzinski commented:
The merge with the logger changes was difficult, dealing with the errors now...

JFRudzinski committed:
fixed utils mypy error

ladinesa commented on src/nomad_simulations/schema_packages/atoms_state.py:
```diff
@@ -569,6 +569,12 @@ class ParticleState(Entity):
         """,
     )
 
+    def get_label(self) -> str:
```
why do we need this wrapper

ladinesa replied:
oh sorry now i get it, but I suggest you make it as an `@abstractmethod` so the children will have to provide this

JFRudzinski replied:
actually, I think it could be needed at the ParticleStates level, perhaps we can discuss this in the system meeting today

JFRudzinski committed:
ruff

JFRudzinski committed:
commented out get_symbols logger for now

JFRudzinski committed:
moved ArchiveSection out of type checking, utils

JFRudzinski committed:
ruff

ladinesa commented on src/nomad_simulations/schema_packages/general.py:
```diff
@@ -206,98 +206,65 @@ def _set_system_branch_depth(
+        if system.composition_formula is not None:
+            return system.composition_formula
+
+        def get_formula_from_particles(system: ModelSystem) -> str | None:
```
maybe get rid of this function as we simplified the logic already

JFRudzinski committed:
clean out old bond_list function

JFRudzinski committed:
remove get_formula from particles

EBB2675 commented:
In the `composition_formula` docstring we say:

>     At the deepest branch in the hierarchy, the composition_formula is expressed in terms of the atomic labels.

Right now a root system that is also a leaf (no `sub_systems`) ends up with `composition_formula = None`

Do we populate `composition_formula` _only when_ the root has subsystems? If yes, we should adjust the docstring

ladinesa commented on src/nomad_simulations/schema_packages/model_system.py:
```diff
@@ -1033,55 +1031,58 @@ class ModelSystem(System):
 
-    @property
-    def _is_atomic(self) -> bool:
+    def is_atomic(self) -> bool:
```
we probably do not need this wrapper just access thr cache and return False by default

JFRudzinski replied:
sorry, I do not get it, can you send some pseudo-code?

ladinesa replied:
perhaps you can  put l1040-43 directly under normalize? and then you can simply access the cache so you do not worry about this being filled in before particle_states are fliled.

JFRudzinski replied:
hmm, interesting option. I have now changed the logic in is_atomic() so that it is consistent with the reassignment logic. Also, the reassignment criterion itself is simplified:
- always reassign particle_states list that are 1. Contain generic ParticleState instances, 2. Are mixed or inconsistent with our is_atomic() logic

I think in this way I have also solved the issue. However, we should still reassess whether what I have done now is good practice. I leave that to #248 as I would like to move on to some other tasks now

ladinesa commented:
From my side you can merge once you resolve remaining minor suggestions.

JFRudzinski committed:
get composition formula for root

JFRudzinski commented:
> In the `composition_formula` docstring we say:
> 
> > ```
> > At the deepest branch in the hierarchy, the composition_formula is expressed in terms of the atomic labels.
> > ```
> 
> Right now a root system that is also a leaf (no `sub_systems`) ends up with `composition_formula = None`
> 
> Do we populate `composition_formula` _only when_ the root has subsystems? If yes, we should adjust the docstring

Fixed so that empty root has composition formula, and added tests

JFRudzinski committed:
new is_atomic criterion, new reassignment policy for ParticleStates

JFRudzinski committed:
one more logical tweak for particle_states reassignment and is_atomic

JFRudzinski committed:
clean are_valid_chemical_symbols() out

JFRudzinski committed:
fixed bug in test
